File: backend/app/services/ai_service.py
"""
AI classification service using Remote AI API.
"""
import httpx
import json
from typing import Dict, Optional, Literal, List
from datetime import datetime

# Constants removed - now loaded from database

# Classification labels
ClassificationLabel = Literal["Interesantes", "SPAM", "EnCopia", "Servicios"]

CLASSIFICATION_PROMPT = """
Eres un asistente de clasificación de correos electrónicos para la empresa Hawkins (@hawkins.es).

**Contexto del correo:**
- De: {from_name} <{from_email}>
- Para: {to_addresses}
- CC: {cc_addresses}
- Asunto: {subject}
- Fecha: {date}
- Cuerpo (primeras 500 palabras): {body_preview}

**Categorías disponibles:**

1. **Interesantes**: Correos con intención real de contratar servicios de Hawkins (presupuestos, propuestas comerciales, reuniones de negocio).

2. **SPAM**: Spam clásico, phishing, newsletters no solicitadas, Y MUY IMPORTANTE: cualquier correo cuyo propósito sea vendernos algo u ofrecernos sus servicios (cold outreach).

3. **EnCopia**: Correos donde hay múltiples destinatarios internos @hawkins.es en To o CC (no dirigidos solo a mí).

4. **Servicios**: Notificaciones transaccionales de plataformas conocidas (booking, bancos, Amazon, etc.).

**IMPORTANTE:**
- Si el correo intenta vendernos algo → SPAM
- Si solicitan nuestros servicios → Interesantes
- Responde SOLO con JSON válido, sin texto adicional.

**Formato de respuesta (JSON estricto):**
{{
  "label": "Interesantes|SPAM|EnCopia|Servicios",
  "confidence": 0.85,
  "rationale": "Máximo 2 frases explicando la decisión"
}}
"""

REVIEW_PROMPT = """
Eres un asistente de clasificación de correos. Dos modelos han clasificado el mismo correo y han llegado a conclusiones diferentes. Debes tomar la decisión final.

**Correo original:**
- De: {from_name} <{from_email}>
- Para: {to_addresses}
- Asunto: {subject}
- Cuerpo: {body_preview}

**Tu clasificación previa:**
- Label: {gpt_label}
- Confianza: {gpt_confidence}
- Razón: {gpt_rationale}

**Clasificación del segundo modelo (Qwen):**
- Label: {qwen_label}
- Confianza: {qwen_confidence}
- Razón: {qwen_rationale}

**Instrucciones:**
Analiza ambas clasificaciones y el correo original. Emite una decisión final única.

**Formato de respuesta (JSON estricto):**
{{
  "final_label": "Interesantes|SPAM|EnCopia|Servicios",
  "final_reason": "Máximo 3 frases explicando tu decisión final",
  "why_not_other": "Máximo 2 frases explicando por qué descartaste la otra clasificación"
}}
"""


class RemoteAIClient:
    """Client for Remote AI API."""
    
    def __init__(self, api_url: str, api_key: str):
        self.api_url = api_url
        self.api_key = api_key
    
    async def generate(
        self,
        model: str,
        prompt: str,
        format: str = "json",
        timeout: float = 120.0
    ) -> Dict:
        """
        Generate completion from Remote AI API.
        
        Args:
            model: Model name
            prompt: Prompt text
            format: Output format (json or text) - kept for compatibility
            timeout: Request timeout in seconds
        
        Returns:
            Dict with response in Ollama-compatible format
        """
        async with httpx.AsyncClient(timeout=timeout, verify=False) as client:
            try:
                # Construct generation URL
                # If api_url ends with /models, strip it to get base
                base_url = self.api_url.replace('/chat/models', '').replace('/models', '').rstrip('/')
                
                # Use /chat/chat endpoint as requested
                generate_url = f"{base_url}/chat/chat"
                
                response = await client.post(
                    generate_url,
                    headers={"x-api-key": self.api_key},
                    json={
                        "modelo": model,
                        "prompt": prompt
                    }
                )
                response.raise_for_status()
                
                # Read raw text first for debugging/fallback
                raw_text = response.text
                
                try:
                    data = response.json()
                    success = data.get("success", True)
                    
                    # Adapt response to Ollama format
                    # /api/chat returns "message": {"content": "..."}
                    message_content = ""
                    if "message" in data and "content" in data["message"]:
                        message_content = data["message"]["content"]
                    elif "choices" in data and isinstance(data["choices"], list) and len(data["choices"]) > 0:
                        # OpenAI format
                        first_choice = data["choices"][0]
                        if "message" in first_choice and "content" in first_choice["message"]:
                            message_content = first_choice["message"]["content"]
                        elif "text" in first_choice:
                            message_content = first_choice["text"]
                    elif "response" in data:
                         # Fallback for /api/generate
                        message_content = data["response"]
                    elif "respuesta" in data:
                        # Fallback for custom API
                        message_content = data["respuesta"]
                    else:
                        # Fallback: keep raw text if we can't extract specific content
                        message_content = raw_text
                
                except json.JSONDecodeError:
                    # Fallback for plain text response (or empty)
                    message_content = raw_text
                    success = True # Assume success if status code was 200
                
                return {
                    "response": message_content,
                    "success": success
                }
            except httpx.HTTPError as e:
                print(f"Remote AI API error: {e}")
                raise
    
    async def check_health(self) -> bool:
        """Check if Remote AI API is accessible."""
        try:
            async with httpx.AsyncClient(timeout=5.0, verify=False) as client:
                response = await client.get(
                    self.api_url,
                    headers={"x-api-key": self.api_key}
                )
                return response.status_code == 200
        except:
            return False
    
    async def list_models(self) -> List[str]:
        """
        Get available models from API.
        Since the API doesn't have a dedicated list endpoint,
        we return common Ollama models.
        """
        # Common Ollama models - user can type custom names too
        common_models = [
            "gpt-oss:120b-cloud",
            "qwen3-coder:480b-cloud",
            "qwen3:latest",
            "llama3:latest",
            "mistral:latest",
            "codellama:latest",
            "gemma:latest",
            "phi3:latest"
        ]
        return common_models


# Legacy alias for compatibility
OllamaClient = RemoteAIClient


# Helper to build prompt
def build_classification_prompt(categories: List[Dict], custom_prompt: Optional[str] = None) -> str:
    """
    Build the classification prompt dynamically based on available categories.
    """
    if custom_prompt:
        # If custom prompt is provided, use it as the base
        base_prompt = custom_prompt
    else:
        # Default prompt structure
        base_prompt = """
Eres un asistente de clasificación de correos electrónicos para la empresa Hawkins (@hawkins.es).

**Contexto del correo:**
- De: {from_name} <{from_email}>
- Para: {to_addresses}
- CC: {cc_addresses}
- Asunto: {subject}
- Fecha: {date}
- Cuerpo (primeras 500 palabras): {body_preview}
"""

    categories_text = "**Categorías disponibles:**\n\n"
    labels_list = []
    
    for idx, cat in enumerate(categories, 1):
        labels_list.append(cat['key'])
        categories_text += f"{idx}. **{cat['key']}**: {cat['ai_instruction']}\n\n"

    labels_joined = "|".join(labels_list)
    
    if custom_prompt:
         # For custom prompt, we just append the categories and JSON instructions if not present?
         # Or we assume the user wrote a full prompt?
         # Best approach: Replace variables in custom prompt if they exist, or append categories.
         # For simplicity: We will expect the user to write {from_name} etc if they want dynamic values.
         # And we will ALWAYS append the JSON instructions at the end to ensure JSON format.
         pass
    else:
        pass # Handle default logic below

    # If using default, build it up
    if not custom_prompt:
        # ... logic as before ...
        pass
    
    # Actually, simplifying:
    # If custom prompt, we just return it formatted with message details + categories + json instructions?
    # No, user prompt should replace the "Contexto" part mainly.
    
    # Revised strategy:
    # If custom_prompt is None, use default.
    # If provided, use it but append categories and JSON constraints to ensure it works.
    
    prompt_text = base_prompt if custom_prompt else base_prompt + categories_text

    instructions_part = f"""
**IMPORTANTE:**
- Clasifica el correo en ÚNICAMENTE una de las categorías anteriores.
- Responde SOLO con JSON válido, sin texto adicional.

**Formato de respuesta (JSON estricto):**
{{{{
  "label": "{labels_joined}",
  "confidence": 0.85,
  "rationale": "Máximo 2 frases explicando la decisión"
}}}}
"""
    return prompt_text + instructions_part


async def classify_with_model(
    message_data: Dict,
    model: str,
    categories: List[Dict],
    ai_client: Optional[RemoteAIClient] = None,
    custom_prompt: Optional[str] = None
) -> Dict:
    """
    Classify message with a specific model.
    """
    if ai_client is None:
        # Load config from DB
        config = await _get_ai_config()
        ai_client = RemoteAIClient(config["api_url"], config["api_key"])
    
    # Prepare prompt
    body_preview = (message_data.get("body_text") or message_data.get("snippet") or "")[:500]
    
    if custom_prompt:
         # Use custom prompt template
         dynamic_prompt_template = build_classification_prompt(categories, custom_prompt)
    else:
         dynamic_prompt_template = build_classification_prompt(categories)
    
    prompt = dynamic_prompt_template.format(
        from_name=message_data.get("from_name", ""),
        from_email=message_data.get("from_email", ""),
        to_addresses=message_data.get("to_addresses", ""),
        cc_addresses=message_data.get("cc_addresses", ""),
        subject=message_data.get("subject", ""),
        date=message_data.get("date", ""),
        body_preview=body_preview
    )
    
    # Call Remote AI
    try:
        response = await ai_client.generate(model, prompt, format="json")
        
        # Parse JSON response
        result_text = response.get("response", "{}")
        classification = json.loads(result_text)
        
        return {
            "label": classification.get("label"),
            "confidence": classification.get("confidence", 0.0),
            "rationale": classification.get("rationale", "")
        }
    
    except Exception as e:
        print(f"Error classifying with {model}: {e}")
        # Return default classification on error
        return {
            "label": "SPAM",
            "confidence": 0.0,
            "rationale": f"Error: {str(e)}"
        }


async def review_with_gpt(
    message_data: Dict,
    gpt_result: Dict,
    qwen_result: Dict,
    ai_client: Optional[RemoteAIClient] = None,
    gpt_model: str = None
) -> Dict:
    """
    GPT reviews both classifications and makes final decision.
    
    Args:
        message_data: Original message data
        gpt_result: GPT's initial classification
        qwen_result: Qwen's classification
        ai_client: Optional RemoteAIClient instance
        gpt_model: Primary model name
    
    Returns:
        Dict with final_label, final_reason, why_not_other
    """
    if ai_client is None:
        config = await _get_ai_config()
        ai_client = RemoteAIClient(config["api_url"], config["api_key"])
        gpt_model = config["primary_model"]
    
    body_preview = (message_data.get("body_text") or message_data.get("snippet") or "")[:500]
    
    prompt = REVIEW_PROMPT.format(
        from_name=message_data.get("from_name", ""),
        from_email=message_data.get("from_email", ""),
        to_addresses=message_data.get("to_addresses", ""),
        subject=message_data.get("subject", ""),
        body_preview=body_preview,
        gpt_label=gpt_result["label"],
        gpt_confidence=gpt_result["confidence"],
        gpt_rationale=gpt_result["rationale"],
        qwen_label=qwen_result["label"],
        qwen_confidence=qwen_result["confidence"],
        qwen_rationale=qwen_result["rationale"]
    )
    
    try:
        response = await ai_client.generate(gpt_model, prompt, format="json")
        result_text = response.get("response", "{}")
        review = json.loads(result_text)
        
        return {
            "final_label": review.get("final_label"),
            "final_reason": review.get("final_reason", ""),
            "why_not_other": review.get("why_not_other", "")
        }
    
    except Exception as e:
        print(f"Error in GPT review: {e}")
        # Default to GPT's original classification
        return {
            "final_label": gpt_result["label"],
            "final_reason": f"Review error, defaulting to GPT: {gpt_result['rationale']}",
            "why_not_other": str(e)
        }


async def _get_ai_config() -> Dict:
    """Load AI configuration from database."""
    try:
        from app.database import AsyncSessionLocal
        from app.models import AIConfig
        from sqlalchemy import select
        
        async with AsyncSessionLocal() as db:
            result = await db.execute(select(AIConfig).limit(1))
            config = result.scalar_one_or_none()
            
            if not config:
                # Return defaults if not found
                print("⚠️ Warning: No AI config found in database, using defaults")
                return {
                    "api_url": "https://192.168.1.45/chat/models",
                    "api_key": "OllamaAPI_2024_K8mN9pQ2rS5tU7vW3xY6zA1bC4eF8hJ0lM",
                    "primary_model": "gpt-oss:120b-cloud",
                    "secondary_model": "qwen3-coder:480b-cloud"
                }
            
            return {
                "api_url": config.api_url,
                "api_key": config.api_key,
                "primary_model": config.primary_model,
                "secondary_model": config.secondary_model
            }
    except Exception as e:
        print(f"❌ Error loading AI config: {e}, using defaults")
        return {
            "api_url": "https://192.168.1.45/chat/models",
            "api_key": "OllamaAPI_2024_K8mN9pQ2rS5tU7vW3xY6zA1bC4eF8hJ0lM",
            "primary_model": "gpt-oss:120b-cloud",
            "secondary_model": "qwen3-coder:480b-cloud"
        }


async def classify_message(message_data: Dict, categories: List[Dict], custom_prompt: Optional[str] = None) -> Dict:
    """
    Full classification pipeline with consensus/tiebreaker.
    
    Args:
        message_data: Dict with message fields
        categories: List of category dicts (key, ai_instruction)
        custom_prompt: Optional custom classification prompt
    
    Returns:
        Dict with complete classification result
    """
    # Load AI config
    config = await _get_ai_config()
    ai_client = RemoteAIClient(config["api_url"], config["api_key"])
    
    # Check if Remote AI is accessible
    if not await ai_client.check_health():
        return {
            "status": "error",
            "error": "Remote AI API is not running or not accessible"
        }
    
    # Step 1: Classify with both models
    gpt_result = await classify_with_model(message_data, config["primary_model"], categories, ai_client, custom_prompt)
    qwen_result = await classify_with_model(message_data, config["secondary_model"], categories, ai_client, custom_prompt)
    
    # Step 2: Check for consensus
    if gpt_result["label"] == qwen_result["label"]:
        # Consensus reached
        return {
            "status": "success",
            "gpt_label": gpt_result["label"],
            "gpt_confidence": gpt_result["confidence"],
            "gpt_rationale": gpt_result["rationale"],
            "qwen_label": qwen_result["label"],
            "qwen_confidence": qwen_result["confidence"],
            "qwen_rationale": qwen_result["rationale"],
            "final_label": gpt_result["label"],
            "final_reason": f"Consensus: {gpt_result['rationale']}",
            "decided_by": "consensus"
        }
    
    else:
        # Disagreement - GPT reviews
        review_result = await review_with_gpt(message_data, gpt_result, qwen_result, ai_client, config["primary_model"])
        
        return {
            "status": "success",
            "gpt_label": gpt_result["label"],
            "gpt_confidence": gpt_result["confidence"],
            "gpt_rationale": gpt_result["rationale"],
            "qwen_label": qwen_result["label"],
            "qwen_confidence": qwen_result["confidence"],
            "qwen_rationale": qwen_result["rationale"],
            "final_label": review_result["final_label"],
            "final_reason": review_result["final_reason"],
            "decided_by": "gpt_review"
        }

REPLY_PROMPT = """
Eres un asistente de redacción de correos electrónicos.
Tu trabajo es redactar una respuesta profesional y cortés a un correo recibido.

**Perfil del Propietario (QUIEN ERES TU):**
{owner_profile}

**Correo Original recibido:**
- De: {from_name} <{from_email}>
- Asunto: {subject}
- Cuerpo: {body_text}

**Instrucción del usuario (si la hay):**
{user_instruction}

**Tu Tarea:**
Redacta ÚNICAMENTE el cuerpo del correo de respuesta. No incluyas "Asunto:" ni saludos/despedidas genéricos si ya están en la firma (aunque un "Hola [Nombre]," al inicio está bien).
Sigue estrictamente el tono y estilo definidos en el perfil del propietario.
"""

async def generate_reply(
    message_data: Dict,
    user_instruction: str = "",
    owner_profile: str = "Eres profesional y eficiente.",
    ai_client: Optional[RemoteAIClient] = None
) -> Dict:
    """
    Generate a reply for a message.
    """
    if ai_client is None:
        config = await _get_ai_config()
        ai_client = RemoteAIClient(config["api_url"], config["api_key"])
        gpt_model = config["primary_model"]
    else:
        config = await _get_ai_config()
        gpt_model = config["primary_model"]

    prompt = REPLY_PROMPT.format(
        from_name=message_data.get("from_name", ""),
        from_email=message_data.get("from_email", ""),
        subject=message_data.get("subject", ""),
        body_text=(message_data.get("body_text") or "")[:1000],
        user_instruction=user_instruction or "Responde adecuadamente al contexto.",
        owner_profile=owner_profile
    )

    # Continue with secure JSON prompt construction defined below
        
    # Re-doing the function properly with JSON for safety
    
    json_prompt = prompt + """
    
    **Formato de respuesta (JSON):**
    {
        "reply_body": "El texto del correo aquí..."
    }
    """
    
    try:
        response = await ai_client.generate(gpt_model, json_prompt, format="json")
        result_text = response.get("response", "{}")
        
        try:
            result_json = json.loads(result_text)
            reply_body = result_json.get("reply_body", "")
            if not reply_body and isinstance(result_json, dict):
                 # Maybe it's directly in another key or just empty?
                 # If keys look like "body", "content", try them
                 reply_body = result_json.get("body") or result_json.get("content") or ""
            
            if not reply_body:
                 # If JSON parsed but empty/irrelevant, fallback to text if it looks like prose
                 # But if result_text was literally "{}", nothing to do.
                 if len(result_text) > 5 and "{" not in result_text[:5]:
                      reply_body = result_text 
        except json.JSONDecodeError:
            # Not JSON, assume it's the text body directly
            reply_body = result_text

        return {
            "reply_body": reply_body
        }
    except Exception as e:
         return {
            "reply_body": f"Error generando respuesta: {str(e)}"
        }

File: app/Services/AiService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Http;
use App\Models\AiConfig;
use Illuminate\Support\Facades\Log;

/**
 * AiService — compatible con la API personalizada en https://192.168.1.45/chat
 *
 * Formato de la API:
 *   POST {api_url}/chat   body: {"prompt":"...","model":"..."}
 *   Header: x-api-key: {api_key}
 *   Respuesta: texto plano o {"response":"..."} o {"choices":[...]}
 *
 *   GET {api_url}/text/models
 *   Respuesta: array de nombres de modelos o {"models":[...]}
 */
class AiService
{
    private ?AiConfig $config = null;

    public function __construct()
    {
        $this->config = AiConfig::first();
    }

    // -------------------------------------------------------------------------
    // Método principal: Clasifica un mensaje
    // -------------------------------------------------------------------------

    /**
     * Clasifica un mensaje usando el modelo primario y secundario.
     */
    public function classifyMessage(array $messageData, array $categories, ?string $customPrompt = null): array
    {
        if (!$this->config) {
            Log::warning('AiService: No hay configuración IA disponible.');
            return $this->fallbackResult();
        }

        $primaryModel   = $this->config->primary_model;
        $secondaryModel = $this->config->secondary_model ?: $primaryModel;

        $prompt = $this->buildClassificationPrompt($messageData, $categories, $customPrompt);

        // Llamamos al modelo primario
        $gptResult = $this->callModel($primaryModel, $prompt);

        // Llamamos al modelo secundario (puede ser el mismo)
        $qwenResult = $primaryModel !== $secondaryModel
            ? $this->callModel($secondaryModel, $prompt)
            : $gptResult;

        $gptLabel  = $gptResult['label']  ?? null;
        $qwenLabel = $qwenResult['label'] ?? null;

        $result = [
            'gpt_label'       => $gptLabel,
            'gpt_confidence'  => $gptResult['confidence']  ?? null,
            'gpt_rationale'   => $gptResult['rationale']   ?? null,
            'qwen_label'      => $qwenLabel,
            'qwen_confidence' => $qwenResult['confidence'] ?? null,
            'qwen_rationale'  => $qwenResult['rationale']  ?? null,
            'final_label'     => null,
            'final_reason'    => null,
            'decided_by'      => null,
            'status'          => 'ok',
        ];

        if (!$gptLabel && !$qwenLabel) {
            Log::warning('AiService: Ambos modelos fallaron.');
            return array_merge($result, $this->fallbackResult());
        }

        if ($gptLabel && !$qwenLabel) {
            $result['final_label']  = $gptLabel;
            $result['final_reason'] = 'Solo el modelo primario respondió.';
            $result['decided_by']   = 'primary';
            return $result;
        }

        if (!$gptLabel && $qwenLabel) {
            $result['final_label']  = $qwenLabel;
            $result['final_reason'] = 'Solo el modelo secundario respondió.';
            $result['decided_by']   = 'secondary';
            return $result;
        }

        // Ambos respondieron
        if (strtolower(trim($gptLabel)) === strtolower(trim($qwenLabel))) {
            $result['final_label']  = $gptLabel;
            $result['final_reason'] = 'Ambos modelos coincidieron.';
            $result['decided_by']   = 'consensus';
            return $result;
        }

        // Desacuerdo: usar primario
        $result['final_label']  = $gptLabel;
        $result['final_reason'] = "Desacuerdo entre modelos ({$gptLabel} vs {$qwenLabel}). Se usó el primario.";
        $result['decided_by']   = 'primary';
        return $result;
    }

    // -------------------------------------------------------------------------
    // Genera respuesta de email
    // -------------------------------------------------------------------------

    public function generateReply(
        string $originalFromName,
        string $originalFromEmail,
        string $originalSubject,
        string $originalBody,
        string $userInstruction,
        string $ownerProfile
    ): string {
        if (!$this->config) {
            throw new \RuntimeException('No hay configuración IA disponible.');
        }

        $prompt = <<<PROMPT
Eres un asistente de redacción de emails profesional.
Perfil del remitente: {$ownerProfile}

Redacta una respuesta al siguiente email siguiendo la instrucción del usuario.
Devuelve SOLO el texto del email de respuesta, sin asunto, sin encabezados, sin explicaciones.
NO uses placeholders ni campos entre corchetes o llaves (ejemplos prohibidos: [Tu nombre], [Nombre], {firma}, {empresa}).
El texto debe quedar final y utilizable tal cual, sin pedir datos adicionales.

Email original:
- De: {$originalFromName} <{$originalFromEmail}>
- Asunto: {$originalSubject}
- Contenido: {$originalBody}

Instrucción: {$userInstruction}

Respuesta:
PROMPT;

        $content = $this->callModelRaw($this->config->primary_model, $prompt);

        if ($content === null) {
            throw new \RuntimeException('La IA no devolvió contenido.');
        }

        return trim($content);
    }

    // -------------------------------------------------------------------------
    // Comprueba disponibilidad de la IA
    // -------------------------------------------------------------------------

    public function checkStatus(): array
    {
        if (!$this->config || !$this->config->api_url || !$this->config->api_key) {
            return ['available' => false, 'reason' => 'Sin configuración IA'];
        }

        try {
            $content = $this->callModelRaw($this->config->primary_model, 'ping', 5);

            if ($content !== null) {
                return ['available' => true, 'model' => $this->config->primary_model];
            }

            return ['available' => false, 'reason' => 'La IA no respondió'];
        } catch (\Throwable $e) {
            return ['available' => false, 'reason' => $e->getMessage()];
        }
    }

    // -------------------------------------------------------------------------
    // Llamadas HTTP internas
    // -------------------------------------------------------------------------

    /**
     * Llama al modelo y devuelve la clasificación parseada.
     */
    private function callModel(string $modelName, string $prompt): array
    {
        try {
            $content = $this->callModelRaw($modelName, $prompt);

            if ($content === null) {
                return ['error' => 'Sin respuesta del modelo'];
            }

            return $this->parseAiResponse($content);
        } catch (\Throwable $e) {
            Log::error("AiService: Error llamando modelo {$modelName}", ['error' => $e->getMessage()]);
            return ['error' => $e->getMessage()];
        }
    }

    /**
     * Llamada HTTP cruda a la API personalizada.
     * Devuelve el texto de respuesta o null si falla.
     */
    private function callModelRaw(string $modelName, string $prompt, int $timeout = 60): ?string
    {
        if (!$this->config) return null;

        $endpoint = $this->resolveChatEndpoint((string)$this->config->api_url);

        $response = Http::withHeaders(['x-api-key' => $this->config->api_key])
            ->timeout($timeout)
            ->post($endpoint, [
                'prompt' => $prompt,
                // Compatibilidad con APIs que esperan "modelo" o "model"
                'modelo' => $modelName,
                'model'  => $modelName,
            ]);

        if ($response->failed()) {
            Log::error("AiService: HTTP {$response->status()} desde {$endpoint}", [
                'model' => $modelName,
                'body'  => mb_substr($response->body(), 0, 500),
            ]);
            return null;
        }

        return $this->extractTextFromResponse($response);
    }

    private function resolveChatEndpoint(string $apiUrl): string
    {
        $url = rtrim($apiUrl, '/');

        if (str_ends_with($url, '/chat/chat')) {
            return $url;
        }

        // Normalizar configuraciones antiguas erróneas como /chat/text/chat
        if (str_ends_with($url, '/chat/text/chat')) {
            return substr($url, 0, -strlen('/text/chat')) . '/chat';
        }

        if (str_ends_with($url, '/chat')) {
            return $url;
        }

        return $url . '/chat/chat';
    }

    /**
     * Extrae el texto de la respuesta, manejando múltiples formatos posibles.
     * Soporta la API personalizada hawkins.es: {"respuesta":"...","success":true}
     */
    private function extractTextFromResponse($response): ?string
    {
        $body = $response->body();

        // Intentar parsear JSON
        $data = $response->json();

        if (is_array($data)) {
            // Formato hawkins.es: {"respuesta": "...", "success": true}
            if (isset($data['respuesta'])) return (string) $data['respuesta'];

            // Formato: {"response": "..."}
            if (isset($data['response'])) return (string) $data['response'];

            // Formato OpenAI: {"choices":[{"message":{"content":"..."}}]}
            if (isset($data['choices'][0]['message']['content'])) {
                return (string) $data['choices'][0]['message']['content'];
            }

            // Formato: {"text": "..."}
            if (isset($data['text'])) return (string) $data['text'];

            // Formato: {"content": "..."}
            if (isset($data['content'])) return (string) $data['content'];

            // Formato: {"message": "..."}
            if (isset($data['message']) && is_string($data['message'])) {
                return (string) $data['message'];
            }

            // Si es un array plano con texto
            if (count($data) === 1) {
                $val = reset($data);
                if (is_string($val)) return $val;
            }
        }

        // Si el body es texto plano directamente
        if (is_string($body) && strlen($body) > 0) {
            return trim($body);
        }

        return null;
    }

    // -------------------------------------------------------------------------
    // Construcción de prompts
    // -------------------------------------------------------------------------

    private function buildClassificationPrompt(array $messageData, array $categories, ?string $customPrompt): string
    {
        $fromName  = $messageData['from_name']  ?? '';
        $fromEmail = $messageData['from_email'] ?? '';
        $subject   = $messageData['subject']    ?? '';
        $date      = $messageData['date']       ?? '';
        $snippet   = $messageData['snippet']    ?? '';
        $bodyText  = mb_substr($messageData['body_text'] ?? '', 0, 1500);

        $toAddresses = $messageData['to_addresses'] ?? [];
        $ccAddresses = $messageData['cc_addresses'] ?? [];

        if (is_string($toAddresses)) $toAddresses = json_decode($toAddresses, true) ?? [];
        if (is_string($ccAddresses)) $ccAddresses = json_decode($ccAddresses, true) ?? [];

        $toList = implode(', ', array_map(fn($a) => is_array($a) ? ($a['email'] ?? '') : $a, $toAddresses));
        $ccList = implode(', ', array_map(fn($a) => is_array($a) ? ($a['email'] ?? '') : $a, $ccAddresses));

        $categoriesText = '';
        foreach ($categories as $cat) {
            $key         = $cat['key']            ?? '';
            $name        = $cat['name']           ?? $key;
            $instruction = $cat['ai_instruction'] ?? '';
            $categoriesText .= "- \"{$name}\" (key: {$key}): {$instruction}\n";
        }

        $customSection = $customPrompt ? "\n## Instrucciones adicionales:\n{$customPrompt}\n" : '';

        return <<<PROMPT
Eres un clasificador de correos electrónicos. Analiza el mensaje y clasifícalo en UNA de las categorías disponibles.

## Categorías disponibles:
{$categoriesText}
{$customSection}
## Datos del mensaje:
- De: {$fromName} <{$fromEmail}>
- Para: {$toList}
- CC: {$ccList}
- Asunto: {$subject}
- Fecha: {$date}
- Resumen: {$snippet}
- Cuerpo:
{$bodyText}

## Instrucciones:
1. Usa el campo "label" con el valor exacto del "key" de la categoría elegida.
2. Asigna una confianza entre 0.0 y 1.0.
3. Proporciona una breve justificación en español.

## Respuesta (SOLO JSON, sin texto adicional):
{"label": "key_de_la_categoria", "confidence": 0.95, "rationale": "Justificación breve en español"}
PROMPT;
    }

    // -------------------------------------------------------------------------
    // Parsing y fallback
    // -------------------------------------------------------------------------

    private function parseAiResponse(string $content): array
    {
        $content = trim($content);

        // Extraer de bloques ```json...```
        if (preg_match('/```(?:json)?\s*([\s\S]+?)\s*```/i', $content, $matches)) {
            $content = trim($matches[1]);
        }

        // Encontrar primer objeto JSON
        if (preg_match('/\{[\s\S]*\}/u', $content, $matches)) {
            $content = $matches[0];
        }

        $data = json_decode($content, true);

        if (json_last_error() !== JSON_ERROR_NONE) {
            Log::warning('AiService: No se pudo parsear JSON', ['content' => mb_substr($content, 0, 300)]);
            return ['error' => 'JSON inválido'];
        }

        $label = $data['label'] ?? null;
        if (!$label) {
            return ['error' => 'Respuesta sin campo "label"'];
        }

        return [
            'label'      => trim((string) $label),
            'confidence' => is_numeric($data['confidence'] ?? null) ? (float) $data['confidence'] : 0.5,
            'rationale'  => isset($data['rationale']) ? (string) $data['rationale'] : '',
        ];
    }

    private function fallbackResult(): array
    {
        return [
            'gpt_label'       => null,
            'gpt_confidence'  => null,
            'gpt_rationale'   => null,
            'qwen_label'      => null,
            'qwen_confidence' => null,
            'qwen_rationale'  => null,
            'final_label'     => 'Servicios',
            'final_reason'    => 'Fallback por error en la IA.',
            'decided_by'      => 'rule_fallback',
            'status'          => 'fallback',
        ];
    }
}
